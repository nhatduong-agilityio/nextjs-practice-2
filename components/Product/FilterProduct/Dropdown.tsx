// Libs
import {
  Button,
  FormControl,
  Input,
  InputGroup,
  InputLeftElement,
  Text,
  useCheckboxGroup,
  VStack,
} from '@chakra-ui/react'
import React, { Fragment, memo, useState } from 'react'

// Components
import ArrowBottom from '@components/Icons/ArrowBottom'
import ArrowRight from '@components/Icons/ArrowRight'
import CheckBox from '@components/Common/CheckBox'
import ButtonLink from '@components/Common/ButtonLink'

type TOption = {
  id: string
  label: string
  value: string | number
}

type TDropdown = {
  id: string
  label: string
  value: string | number
}

type TNestedItems = {
  id: string
  label: string
  value: TOption[][]
}

interface IProps {
  name: string
  items?: TDropdown[]
  type: 'checkBox' | 'button' | 'input'
  nestedItems?: TNestedItems[]
  level?: '1' | '2'
  search: string | undefined
  selectedCategory?: string[]
}

const Dropdown = memo(
  ({ name, items, type, nestedItems, level = '1', search, selectedCategory }: IProps) => {
    const [open, setOpen] = useState(false)

    // TODO: add value to get data from check box
    const { getCheckboxProps } = useCheckboxGroup({
      defaultValue: [],
    })

    const handleOpen = () => {
      setOpen(!open)
    }

    // console.log(value.sort().join(' and '))

    const renderButton = () =>
      level === '1' ? (
        <Button
          variant='ghost'
          justifyContent='space-between'
          height='fit-content'
          style={{
            paddingInline: 0,
          }}
          fontSize={['16px', '18px']}
          rightIcon={
            open ? (
              <ArrowBottom width='13px' height='8px' viewBox='0 0 13 8' fill='#29293E' />
            ) : (
              <ArrowRight width='8px' height='13px' viewBox='0 0 8 13' fill='#29293E' />
            )
          }
          onClick={handleOpen}
        >
          {name.toLowerCase()}
        </Button>
      ) : (
        <Button
          w='full'
          variant='ghost'
          fontWeight='Regular'
          justifyContent='start'
          height='fit-content'
          style={{
            paddingInline: 0,
          }}
          fontSize={['16px', '18px']}
          leftIcon={
            open ? (
              <ArrowBottom
                style={{ marginRight: '8px' }}
                width='13px'
                height='8px'
                viewBox='0 0 13 8'
                fill='#29293E'
              />
            ) : (
              <ArrowRight
                style={{ marginRight: '13px' }}
                width='8px'
                height='13px'
                viewBox='0 0 8 13'
                fill='#29293E'
              />
            )
          }
          onClick={handleOpen}
        >
          {name.toLowerCase()}
        </Button>
      )

    return (
      <VStack style={{ margin: 0 }} w='full' pt='15px'>
        {renderButton()}

        <VStack style={{ margin: 0 }} w='full' display={open ? 'flex' : 'none'}>
          {type === 'checkBox' &&
            items &&
            items.map((item) => (
              <Fragment key={item.id}>
                <CheckBox {...getCheckboxProps({ name: item.label, value: item.value })} />
              </Fragment>
            ))}
          {type === 'button' &&
            items &&
            items.map((item) => (
              <Fragment key={item.id}>
                <ButtonLink
                  name={item.value}
                  href={{
                    pathname: '/product',
                    query: {
                      ...(search && { search }),
                      ...(item.value !== 'All' && { category: item.value }),
                    },
                  }}
                  active={selectedCategory?.includes(item.value.toString())}
                />
              </Fragment>
            ))}
          {type === 'input' && items && (
            <form style={{ marginTop: 0, width: '100%' }}>
              {items.map((item) => (
                <Fragment key={item.id}>
                  <FormControl style={{ margin: 0 }} pt='10px'>
                    <InputGroup>
                      <InputLeftElement
                        w='56px'
                        pointerEvents='none'
                        borderLeftRadius='5px'
                        borderY='solid 1px #CACACA'
                        borderLeft='solid 1px #CACACA'
                        bgColor='#E6E6E6'
                      >
                        <Text as='span' size='20' variant='small'>
                          {name}
                        </Text>
                      </InputLeftElement>
                      <Input
                        type='number'
                        placeholder={item.label || ''}
                        variant='outline'
                        pl='76px'
                        borderRadius='5px'
                      />
                    </InputGroup>
                  </FormControl>
                </Fragment>
              ))}
            </form>
          )}
          {nestedItems &&
            nestedItems.map((item, index) => (
              <Fragment key={item.label}>
                <Dropdown
                  name={item.label}
                  items={item.value[index]}
                  type='button'
                  level='2'
                  search={search}
                  selectedCategory={selectedCategory}
                />
              </Fragment>
            ))}
        </VStack>
      </VStack>
    )
  },
)

Dropdown.displayName = 'Dropdown'

export default Dropdown
