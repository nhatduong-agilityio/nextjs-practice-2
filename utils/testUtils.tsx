import { ChakraProvider } from '@chakra-ui/react'
import { AuthProvider } from '@store/providers/AuthProvider'
import { LoadingProvider } from '@store/providers/LoadingProvider'
import type { RenderOptions, RenderResult } from '@testing-library/react'
import { render } from '@testing-library/react'
import { customTheme } from '@themes/index'
import type { ReactElement } from 'react'

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ChakraProvider theme={customTheme}>
      <LoadingProvider>
        <AuthProvider>{children}</AuthProvider>
      </LoadingProvider>
    </ChakraProvider>
  )
}

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>): RenderResult =>
  render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
