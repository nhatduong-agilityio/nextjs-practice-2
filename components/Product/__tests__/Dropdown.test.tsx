// Utils
import { render } from '@utils/testUtils'

import { ColorMock } from '@mock/dataMock'

// Components
import Dropdown from '../FilterProduct/Dropdown'

jest.mock('next/navigation', () => ({
  useSearchParams: jest.fn().mockReturnValue({
    get: jest.fn(),
  }),
}))

describe('Dropdown render', () => {
  const propsMock = {
    name: 'Color',
    items: ColorMock,
    type: 'checkBox' as 'checkBox' | 'button' | 'input',
    search: 'search',
    selectedCategory: ['Chair', 'Sofa'],
  }

  it('Should FilterProduct match Snapshot', () => {
    const { container } = render(<Dropdown {...propsMock} />)

    expect(container).toMatchSnapshot()
  })
})
