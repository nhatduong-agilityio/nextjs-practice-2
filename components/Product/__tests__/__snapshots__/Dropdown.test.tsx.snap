// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown render Should FilterProduct match Snapshot 1`] = `
<div>
  <div
    class="chakra-stack css-1v4uoso"
    style="margin: 0px;"
  >
    <button
      class="chakra-button css-l43bjg"
      style="padding-inline: 0;"
      type="button"
    >
      color
      <span
        class="chakra-button__icon css-1hzyiq5"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-pfz1im"
          focusable="false"
          viewBox="0 0 8 13"
        >
          <path
            d="M0.612558 0.868804C0.68513 0.796049 0.771341 0.738325 0.866256 0.69894C0.96117 0.659555 1.06292 0.639282 1.16568 0.639282C1.26845 0.639282 1.3702 0.659555 1.46511 0.69894C1.56003 0.738325 1.64624 0.796049 1.71881 0.868804L6.91256 6.06255C6.9705 6.12038 7.01647 6.18906 7.04783 6.26466C7.07919 6.34027 7.09534 6.42132 7.09534 6.50318C7.09534 6.58503 7.07919 6.66609 7.04783 6.74169C7.01647 6.8173 6.9705 6.88598 6.91256 6.9438L1.71881 12.1376C1.41256 12.4438 0.918808 12.4438 0.612558 12.1376C0.306308 11.8313 0.306308 11.3376 0.612558 11.0313L5.13756 6.50005L0.606309 1.9688C0.306309 1.6688 0.306308 1.1688 0.612558 0.868804Z"
          />
        </svg>
      </span>
    </button>
    <div
      class="chakra-stack css-19pl6ik"
      style="margin: 0px;"
    >
      <label
        class="css-d3wqq2"
      >
        <input
          hidden=""
          name="Blue"
          style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
          type="checkbox"
          value="blue"
        />
        <div
          aria-hidden="true"
          class="css-i2vche"
        />
        <p
          class="chakra-text css-5l0nl"
        >
          Blue
        </p>
      </label>
      <label
        class="css-d3wqq2"
      >
        <input
          hidden=""
          name="Purple"
          style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
          type="checkbox"
          value="purple"
        />
        <div
          aria-hidden="true"
          class="css-i2vche"
        />
        <p
          class="chakra-text css-5l0nl"
        >
          Purple
        </p>
      </label>
      <label
        class="css-d3wqq2"
      >
        <input
          hidden=""
          name="Red"
          style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
          type="checkbox"
          value="red"
        />
        <div
          aria-hidden="true"
          class="css-i2vche"
        />
        <p
          class="chakra-text css-5l0nl"
        >
          Red
        </p>
      </label>
      <label
        class="css-d3wqq2"
      >
        <input
          hidden=""
          name="Orange"
          style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
          type="checkbox"
          value="orange"
        />
        <div
          aria-hidden="true"
          class="css-i2vche"
        />
        <p
          class="chakra-text css-5l0nl"
        >
          Orange
        </p>
      </label>
      <label
        class="css-d3wqq2"
      >
        <input
          hidden=""
          name="Yellow"
          style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
          type="checkbox"
          value="yellow"
        />
        <div
          aria-hidden="true"
          class="css-i2vche"
        />
        <p
          class="chakra-text css-5l0nl"
        >
          Yellow
        </p>
      </label>
      <label
        class="css-d3wqq2"
      >
        <input
          hidden=""
          name="Green"
          style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
          type="checkbox"
          value="green"
        />
        <div
          aria-hidden="true"
          class="css-i2vche"
        />
        <p
          class="chakra-text css-5l0nl"
        >
          Green
        </p>
      </label>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
